cmd/nerdctl/login.go

```diff
@@ -95,5 +96,10 @@ func verifyLoginOptions(cmd *cobra.Command, options *loginOptions) error {
 		options.password = strings.TrimSuffix(string(contents), "\n")
 		options.password = strings.TrimSuffix(options.password, "\r")
 	}
+
+	if strings.ReplaceAll(url.User(options.username).String(), ",", "") != options.username {
```

Comment:
Sure, I'll replace it asap!

You are right `url.User()` puts our username into a structs, however the only reason I think we need this is the `.String()` method which encodes that username (and password if entered) to RFC 7617 compatible Basic Auth credentials, so basically it does the whole validation step for us.

Reply:
Because of the above mentioned stuff this approach compares a properly encoded user with the plain one. https://go.dev/play/p/QMI-3OARwVk

Reply:
Thanks for clarifying! And thanks for your patience!

Please correct me again if I am wrong below :-)

a. We support http basic auth, but also other forms of authentication that do not have the same restrictions - right now, there is unfortunately no way to know which authentication we are dealing with at that point in the code... we need to be careful what restrictions we are going to impose globally here.

b. it does not seem to me like `url.User` is what should be used for basic authentication. If you look at the go library in the section that does deal with basic authentication, they just base64 encode it without escaping: https://github.com/golang/go/blob/master/src/net/http/client.go#L424 (because that is what the RFC is mandating)
Escaping as provided by `UserInfo.String` AFAIK is meant solely to be used when the user credentials are going to be embedded in the url (in the form of `scheme://user:password@domain`).
Put otherwise, "," is a legit username according the the RFC:
```
  For the user-id, recipients MUST support all characters defined in
   the "UsernameCasePreserved" profile defined in Section 3.3 of
   [RFC7613], with the exception of the colon (":") character.
```

What are your thoughts?

Reply:
FWIW, it looks to me like a username is valid is it is a PRECIS Identifier (https://datatracker.ietf.org/doc/html/draft-ietf-precis-7564bis-10#page-9), which specifically allows (among other things) code points in the range U+0021 to U+007E.

Reply:
I guess I am back to my suggestion that we should blacklist only the colon (in light of the RFC paragraph I quoted).

The unresolved question IMO is point a. above - is it acceptable to forbid colons for usernames for all types of authentication, or are we going to break some folks workflows where they use (for example) TokenAuth and usernames with colons in them (who does that, really? :D :D :D)

Reply:
RFC 7519 states JWTs must be URL safe which means if they contain any colons or they should be encoded, plus it consist of 3 base64 encoded parts which can't contain any colons neither, so with JWTs we are gucci, I'm not so sure about the other token standards.